Firestore/Protos/build_protos.py

```diff
@@ -0,0 +1,204 @@
+    generator_bin = os.path.join(nanopb_dir, 'generator/protoc-gen-nanopb')
+    nanopb_out = 'nanopb'
+
+    nanopb_py = os.path.join(nanopb_dir, 'generator/proto/nanopb_pb2.py')
+    if not os.path.isfile(nanopb_py):
+      self.__download()
+      self.__build(nanopb_dir)
+
+    self.__run_generator(generator_bin, nanopb_out)
+
+    sources = collect_files(nanopb_out, '.nanopb.h', '.nanopb.c')
+    post_process_files(sources, add_copyright, nanopb_rename_delete)
+
+  def __download(self):
+    """Downloads and unpacks nanopb sources."""
+
+    url = 'https://github.com/nanopb/nanopb/archive/%s.tar.gz' % NANOPB_VERSION
+    tgz = 'nanopb-%s.tar.gz' % NANOPB_VERSION
+
+    if not os.path.isfile(tgz):
+      print('Downloading %s' % url)
+      response = urllib2.urlopen(url)
+      with open(tgz, 'wb') as fd:
+        shutil.copyfileobj(response, fd)
+
+    with tarfile.open(tgz) as tar:
+      tar.extractall()
+
+  def __build(self, nanopb_dir):
+    """Builds the nanopb plugin from sources."""
+
+    print('Building %s' % nanopb_dir)
+    cwd = os.getcwd()
+    os.chdir(nanopb_dir)
+
+    subprocess.call(['cmake', '.'])
+    subprocess.call(['make'])
+
+    # Copy built files into place where the generator expects to find them
+    for src in ['plugin_pb2.py', 'nanopb_pb2.py']:
+      dest = os.path.join('generator/proto', src)
+      shutil.copyfile(src, dest)
+
+    os.chdir(cwd)
+
+  def __run_generator(self, generator_bin, out_dir):
+    """Invokes protoc using the nanopb plugin."""
+    nanopb_flags = ' '.join([
+        '--extension=.nanopb',
+        '--options-file=protos/%s.options',
+    ])
+
+    cmd = [
+        PROTOC_BIN,
+        '-I', 'protos',
+        '--plugin=' + generator_bin,
+        '--nanopb_out=%s:%s' % (nanopb_flags, out_dir),
+    ]
+    cmd.extend(self.proto_files)
+
+    subprocess.call(cmd)
+
+
+def post_process_files(filenames, *processors):
+  for filename in filenames:
+    lines = []
+    with open(filename, 'r') as fd:
+      lines = fd.readlines()
+
+    for processor in processors:
+      lines = processor(lines)
+
+    write_file(filename, lines)
+
+
+def write_file(filename, lines):
+  with open(filename, 'w') as fd:
+    fd.write(''.join(lines))
+
+
+def add_copyright(lines):
+  """Adds a copyright notice to the lines."""
+  result = [COPYRIGHT_NOTICE, '\n']
+  result.extend(lines)
+  return result
+
+
+def nanopb_rename_delete(lines):
+  """Renames a delete symbol to delete_.
+
+  If a proto uses a field named 'delete', nanopb happily uses that in the
+  message definition. Works fine for C; not so much for C++.
+
+  Args:
+    lines: The lines to fix.
+
+  Returns:
+    The lines, fixed.
+  """
+  delete_keyword = re.compile(r'\bdelete\b')
+  return [delete_keyword.sub('delete_', line) for line in lines]
+
+
+def collect_files(root_dir, *extensions):
+  """Finds files with the given extensions in the root_dir."""
+  result = []
+  for root, dirs, files in os.walk(root_dir):
+    del dirs
```

Comment:
This is the [style-guide preferred](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#214-decision) way of marking an argument unused (though I forgot to include the comment why I was deleting it). Fixed.

The underscore thing is apparently no longer recommended, though the rationale presented seems to only apply to function arguments rather than return values. I'm fine either way--I'm just appeasing the linter.

Reply:
Yeah, I think the rational is pretty weak in this case; personally, I'd go with an underscore here. It's more concise.

But (obviously) I don't object to following the style guide either.

Reply:
Went with underscore.